frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/Model.java

```diff
@@ -116,19 +122,49 @@ public void addFirst(Job job) {
         jobsDb.get(DEFAULT_DATA_QUEUE).addFirst(job);
     }
 
-    public Job nextJob(String threadId) throws InterruptedException {
-        return nextJob(threadId, Long.MAX_VALUE);
-    }
+    public void pollBatch(String threadId, long waitTime, Map<String, Job> jobsRepo)
+            throws InterruptedException {
+        if (jobsRepo == null || threadId == null) {
+            logger.error("Invalid input given provided");
+            throw new IllegalArgumentException();
+        }
+
+        if (!jobsRepo.isEmpty()) {
+            logger.error("The jobs repo provided contains stale jobs. Clear them!!");
+            throw new IllegalArgumentException();
+        }
 
-    public Job nextJob(String threadId, long timeoutMillis) throws InterruptedException {
-        LinkedBlockingDeque<Job> jobs = jobsDb.get(threadId);
-        if (jobs != null && !jobs.isEmpty()) {
-            Job j = jobs.poll();
+        LinkedBlockingDeque<Job> jobsQueue = jobsDb.get(threadId);
+        if (jobsQueue != null && !jobsQueue.isEmpty()) {
+            Job j = jobsQueue.poll(waitTime, TimeUnit.MILLISECONDS);
             if (j != null) {
-                return j;
+                jobsRepo.put(j.getJobId(), j);
+                return;
+            }
+        }
+
+        synchronized (this) {
+            long maxDelay = maxBatchDelay;
+            long begin = System.currentTimeMillis();
+            jobsQueue = jobsDb.get(DEFAULT_DATA_QUEUE);
+
+            Job j = jobsQueue.poll(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+            logger.trace("get first job: {}", Objects.requireNonNull(j).getJobId());
+
+            jobsRepo.put(j.getJobId(), j);
+            for (int i = 0; i < batchSize - 1; ++i) {
```

Comment:
Any reason for  ```i < batchSize - 1``` and not ```i < batchSize``` ?

Reply:
Good catch. This seems wrong. It doesn't affect a running system, but this reduces the batch size that can be handled by the model by 1.

Reply:
This should be i < batchSize -1, because the first job is inserted into the repo before we enter the job loop. This is because we have a fully-blocking wait until we receive the first message.